### PR TITLE
[GlobalIsel]  Combine selects with constants

### PR DESCRIPTION
**Pull request by tschuett**
A first small step at combining selects.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/andcompare.ll [hunk 1]
```diff
@@ -2514,7 +2517,8 @@ define i32 @cmp_to_ands6(i32 %num) {
 ; GISEL:       // %bb.0:
 ; GISEL-NEXT:    and w8, w0, #0xfe
 ; GISEL-NEXT:    cmp w8, #16
-; GISEL-NEXT:    csel w0, w8, wzr, hs
+; GISEL-NEXT:    cset w9, hs
+; GISEL-NEXT:    and w0, w9, w8
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/andcompare.ll:2521]: This looks worse?

File: llvm/test/CodeGen/AArch64/neon-bitwise-instructions.ll [hunk 2]
```diff
@@ -1100,9 +1100,7 @@ define <8 x i8> @vselect_constant_cond_zero_v8i8(<8 x i8> %a) {
 ; CHECK-GI:       // %bb.0:
 ; CHECK-GI-NEXT:    adrp x8, .LCPI83_0
 ; CHECK-GI-NEXT:    ldr d1, [x8, :lo12:.LCPI83_0]
-; CHECK-GI-NEXT:    shl v1.8b, v1.8b, #7
-; CHECK-GI-NEXT:    sshr v1.8b, v1.8b, #7
-; CHECK-GI-NEXT:    and v0.8b, v0.8b, v1.8b
+; CHECK-GI-NEXT:    and v0.8b, v1.8b, v0.8b
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/neon-bitwise-instructions.ll:1103]: Here are improvements.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/neon-bitwise-instructions.ll:1103]: It is an interplay between this combiner and the imported code. This combiner is too weak. Trying to make it more powerful will always be a compromise. Once it get's more powerful there will be less compromises. Never going to touch this combiner again will be a sad state.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -6318,3 +6262,308 @@ void CombinerHelper::applyCommuteBinOpOperands(MachineInstr &MI) {
...
+      B.buildOr(Dest, Inner, True, Flags);
+    };
+    return true;
+  }
+
+  return false;
+}
+
+// TODO: use knownbits to determine zeros
+bool CombinerHelper::tryFoldBoolSelectToLogic(GSelect *Select,
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6482]: This is a port of foldBoolSelectToLogic from the DAG.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 4]
```diff
@@ -6318,3 +6262,308 @@ void CombinerHelper::applyCommuteBinOpOperands(MachineInstr &MI) {
...
+    if (!Value)
+      Value = IConstant->Value;
+    else if (*Value != IConstant->Value)
+      return std::nullopt;
+  }
+  return Value;
+}
+
+// TODO: use knownbits to determine zeros
+bool CombinerHelper::tryFoldSelectOfConstants(GSelect *Select,
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6345]: This is a partial port of foldSelectOfConstants from the DAG.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fold-binop-into-select.mir [hunk 5]
```diff
@@ -749,8 +751,7 @@ body: |
     ; CHECK-NEXT: %reg:_(s32) = COPY $vgpr0
     ; CHECK-NEXT: %zero:_(s32) = G_CONSTANT i32 0
     ; CHECK-NEXT: %cond:_(s1) = G_ICMP intpred(eq), %reg(s32), %zero
-    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
-    ; CHECK-NEXT: %srem:_(s32) = G_SELECT %cond(s1), [[C]], %zero
+    ; CHECK-NEXT: %srem:_(s32) = G_ZEXT %cond(s1)
```
Comment (tschuett) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fold-binop-into-select.mir:754]: improvement

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fold-binop-into-select.mir [hunk 6]
```diff
@@ -802,8 +803,7 @@ body: |
     ; CHECK-NEXT: %reg:_(s32) = COPY $vgpr0
     ; CHECK-NEXT: %zero:_(s32) = G_CONSTANT i32 0
     ; CHECK-NEXT: %cond:_(s1) = G_ICMP intpred(eq), %reg(s32), %zero
-    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
-    ; CHECK-NEXT: %udiv:_(s32) = G_SELECT %cond(s1), [[C]], %zero
+    ; CHECK-NEXT: %udiv:_(s32) = G_ZEXT %cond(s1)
```
Comment (tschuett) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fold-binop-into-select.mir:806]: improvement

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 7]
```diff
@@ -6352,16 +6352,22 @@ bool CombinerHelper::tryFoldSelectOfConstants(GSelect *Select,
   LLT CondTy = MRI.getType(Select->getCondReg());
   LLT TrueTy = MRI.getType(Select->getTrueReg());
 
-  // Either both are scalars or both are vectors.
-  std::optional<APInt> TrueOpt = getConstantOrConstantSplatVector(True);
-  std::optional<APInt> FalseOpt = getConstantOrConstantSplatVector(False);
+  // Only do this before legalization to avoid conflicting with target-specific
+  // transforms in the other direction.
+  if (CondTy != LLT::scalar(1))
+    return false;
```
Comment (aemerson) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6358]: To implement your comment this check only works for AArch64, you should use `isPreLegalize()` instead.

However I think you actually want what your code does here, i.e. the `s1` check. So probably rewrite the comment to remove the part about before-legalization. For non-s1 types you might need to care about getBooleanContents() etc so easier to do it for s1 only.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 8]
```diff
@@ -6352,16 +6352,22 @@ bool CombinerHelper::tryFoldSelectOfConstants(GSelect *Select,
...
+
+  // Noth are scalars.
```
Comment (aemerson) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6360]: typo "Noth"

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 7 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp]: The text was inspired by the DAG combiner. I updated it.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 9]
```diff
@@ -6318,3 +6262,300 @@ void CombinerHelper::applyCommuteBinOpOperands(MachineInstr &MI) {
...
+      B.buildShl(Dest, Inner, ShAmtC, Flags);
+    };
+    return true;
+  }
+  // select Cond, -1, C --> or (sext Cond), C
+  if (TrueValue.isAllOnes()) {
+    MatchInfo = [=](MachineIRBuilder &B) {
+      B.setInstrAndDebugLoc(*Select);
+      Register Inner = MRI.createGenericVirtualRegister(TrueTy);
+      B.buildSExtOrTrunc(Inner, Cond);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6451]: Can fold this into the MIRBuilder,

auto Inner = B.buildSextOrTrunc(TrueTy, Cond)

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6451]: I prefer the verbose version. It is more readable. I read code more often than I write code. It is completely obvious to any reader that I need a temporary register. Feeding the LLT into the builder is shorter, but it is confusing.

Comment (aemerson) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6451]: It's idiomatic for GlobalISel code to directly pass LLT to the builder when possible, you'll notice that we tend to rarely directly call `createGenericVirtualRegister()`. That said, if you feel very strongly about this I won't block this patch on that basis, but you should also be aware that this code will be in a minority w.r.t the rest of the codebase.

**Review by aemerson — APPROVED** [review]
Approved, I'll leave the final call to you with the coding style.

**Pull request merged**